### PR TITLE
Support to watch/own object not known when the operator starts, but only after a component has been enabled.

### PR DESCRIPTION
**Pull request by lburgazzoli**
Signed-off-by: Luca Burgazzoli <lburgazzoli@gmail.com>

## Description

https://issues.redhat.com/browse/RHOAIENG-15170

## Merge criteria

- [X] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [X] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [X] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [X] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [X] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: controllers/components/dashboard/dashboard_controller.go [hunk 1]
```diff
@@ -57,10 +57,16 @@ func (s *componentHandler) NewComponentReconciler(ctx context.Context, mgr ctrl.
...
+		Owns(&appsv1.Deployment{}, reconciler.WithPredicates(resources.NewDeploymentPredicate())).
 		// operands - openshift
 		Owns(&routev1.Route{}).
 		Owns(&consolev1.ConsoleLink{}).
+		// Those APIs are provided by the component itself hence they should
+		// be watched dynamically
+		OwnsGVK(gvk.AcceleratorProfile, reconciler.Dynamic()).
+		OwnsGVK(gvk.OdhApplication, reconciler.Dynamic()).
+		OwnsGVK(gvk.OdhDocument, reconciler.Dynamic()).
+		OwnsGVK(gvk.OdhQuickStart, reconciler.Dynamic()).
```
Comment (VaishnaviHire) [inline controllers/components/dashboard/dashboard_controller.go:69]: Should we watch `OdhDashboardConfig` for just for delete events?

Comment (lburgazzoli) [reply controllers/components/dashboard/dashboard_controller.go:69]: good point, I'll add it

**Comment by VaishnaviHire** [comment]
@lburgazzoli Is there a way we can test this?

**Comment by lburgazzoli** [comment]
> @lburgazzoli Is there a way we can test this?

there is a `validateOperandsDynamicallyWatchedResources` test for modelmesh which ensure that a dynamically watched resources is reverted in case it gets changed. I'm adding a similar one for the dashboard

We may also add some test based on the metrics endpoint since the number of dynamically watched resources is tracked.

File: controllers/components/dashboard/dashboard_controller.go (hunk 1 above)
Comment (lburgazzoli) [reply controllers/components/dashboard/dashboard_controller.go:70]: done

**Review by VaishnaviHire — APPROVED** [review]
/lgtm

**Review by zdtsw — APPROVED** [review]

**Pull request merged**